File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/csharp/ObsoletePatterns.cs
using System;
using System.Windows.Forms;

namespace ClipboardExamples
{
    // Examples of obsolete patterns that no longer work in .NET 10
    public class ObsoletePatterns
    {
        // <ObsoleteCustomType>
        [Serializable]
        public class Person
        {
            public string Name { get; set; }
            public int Age { get; set; }
        }

        public static void BrokenCustomTypeExample()
        {
            // This worked in .NET 8 and earlier but silently fails starting with .NET 9
            Person person = new Person { Name = "John", Age = 30 };
            Clipboard.SetData("MyApp.Person", person);  // No data is stored

            // Later attempts to retrieve the data return null
            object data = Clipboard.GetData("MyApp.Person");
        }
        // </ObsoleteCustomType>

        // <ObsoleteGetData>
        public static void ObsoleteGetDataExample()
        {
            // Don't use - GetData() is obsolete in .NET 10
            object data = Clipboard.GetData("MyApp.Person");  // Obsolete method

            // Always returns null on a custom object type
            if (data != null)
            {
                Person person = (Person)data;  // Unsafe casting
                ProcessPerson(person);
            }
        }
        // </ObsoleteGetData>

        private static void ProcessPerson(Person person)
        {
            Console.WriteLine($"Processing person: {person.Name}, Age: {person.Age}");
        }
    }
}

File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/vb/ObsoletePatterns.vb
Imports System
Imports System.Windows.Forms

Namespace ClipboardExamples
    ' Examples of obsolete patterns that no longer work in .NET 10
    Public Class ObsoletePatterns
        ' <ObsoleteCustomType>
        <Serializable>
        Public Class Person
            Public Property Name As String
            Public Property Age As Integer
        End Class

        Public Shared Sub BrokenCustomTypeExample()
            ' This worked in .NET 8 and earlier but silently fails starting with .NET 9
            Dim person As New Person With {.Name = "John", .Age = 30}
            Clipboard.SetData("MyApp.Person", person)  ' No data is stored

            ' Later attempts to retrieve the data return a NotSupportedException instance
            Dim data As Object = Clipboard.GetData("MyApp.Person")
        End Sub
        ' </ObsoleteCustomType>

        ' <ObsoleteGetData>
        Public Shared Sub ObsoleteGetDataExample()
            ' Don't use - GetData() is obsolete in .NET 10
            Dim data As Object = Clipboard.GetData("MyApp.Person")  ' Obsolete method

            ' Returns a NotSupportedException instance for a custom object type
            If data IsNot Nothing Then
                Dim person As Person = CType(data, Person)
                Console.WriteLine($"Processing person: {person.Name}, Age: {person.Age}")
            End If
        End Sub
        ' </ObsoleteGetData>
    End Class
End Namespace
